C7Engine/EntryPoints/TurnHandling.cs

```diff
@@ -69,28 +69,33 @@ public static void EndTurn()
 						city.location.unitsOnTile.Add(newUnit);
 						gameData.mapUnits.Add(newUnit);
 						city.owner.AddUnit(newUnit);
-	                }
-					
+					}
+
 					city.SetItemBeingProduced(CityProductionAI.GetNextItemToBeProduced(city, producedItem));
-                }
-            }
-            //Reset movement points available for all units
-            foreach (MapUnit mapUnit in gameData.mapUnits)
-            {
-                mapUnit.movementPointsRemaining = mapUnit.unitType.movement;
-            }
-            //Clear all wait queue, so if a player ended the turn without handling all waited units, they are selected
-            //at the same place in the order.  Confirmed this is what Civ3 does.
-            UnitInteractions.ClearWaitQueue();
-            gameData.turn++;
+				}
+			}
+			//Reset movement points available for all units
+			foreach (MapUnit mapUnit in gameData.mapUnits)
+			{
+				mapUnit.movementPointsRemaining = mapUnit.unitType.movement;
+			}
+			//Clear all wait queue, so if a player ended the turn without handling all waited units, they are selected
+			//at the same place in the order.  Confirmed this is what Civ3 does.
+			UnitInteractions.ClearWaitQueue();
+			gameData.turn++;
 
-	    new MsgStartTurn().send();
-        }
+			foreach (MapUnit mapUnit in gameData.mapUnits) {
```

Comment:
Because this is after the movement reset on lines 78-80, if you give a unit a go to order, it effectively gets double movement for the first turn of its path.

I think this should probably be at the start of the "EndTurn" method.  It should probably also use the gameData.players collection (which has a _units_ list of MapUnits), to only apply this to human players.  It works out to be the same right now, but if the AI starts giving units go to orders, that distinction will be useful.

(For full reference, in Civ3, units with go to orders are actually moved _before_ the turn ends.  All units without a go to order, and which need orders, are cycled through first.  Then units with Go To orders move.  Then, if any units received a "wait" order on their initial cycle, they are activated again.  At this point I don't think we need to perfectly emulate that, and there are times when I don't even like that methodology, particularly the go to happening before the "waited" units.  But at some point we should give that as an option at least)

Reply:
yeah good catch, not sure how I didn't notice any extra movement when testing this. And I agree this could go in an AI. I need to play around with it in code, but I think most behaviour can and probably should be modelled that way since it will prevent pieces of engine code from getting coupled together